### DISCUSSION TITLE
Is there any way to make the Milvus GPU specific Image run on clusters without GPU ?

### DISCUSSION THREAD

**Discussion by Ald392** [milvus-io/milvus#36695] [Q&A and General discussion]
I want to create a single Milvus image that can run on both GPU-enabled and GPU-disabled Kubernetes clusters.

- In GPU-enabled clusters: The image should utilize the GPU.
- In GPU-disabled clusters: The image should default to non-GPU-specific features.

Currently, when I create a Milvus image with GPU support, it works only on clusters with CUDA drivers installed. On clusters without GPU drivers, I receive an error stating "CUDA Driver not found."
Is there a way to achieve the desired behavior, where the image adapts to the cluster's GPU availability?

**Comment by xiaofan-luan** [comment]
you can do it.

You need to run with a gpu image and you can control whether to use GPU with index type.

**Reply by Ald392** [reply]
But when i update the image of my nodes from milvus CPU version to GPU version in a cluster with no GPU drivers nor GPU, the querynode and indexnode crashes with the following logs : 
```
[2024/10/08 10:02:22.332 +00:00] [INFO] [indexnode/indexnode.go:214] ["IndexNode init session successful"] [serverID=66]
WARNING: Logging before InitGoogleLogging() is written to STDERR
I20241008 10:02:22.332165    86 knowhere_config.cc:95] [KNOWHERE][SetBlasThreshold][milvus] Set faiss::distance_compute_blas_threshold to 16384
I20241008 10:02:22.332331    86 knowhere_config.cc:106] [KNOWHERE][SetEarlyStopThreshold][milvus] Set faiss::early_stop_threshold to 0
I20241008 10:02:22.332366    86 knowhere_config.cc:54] [KNOWHERE][ShowVersion][milvus] Knowhere Version: v2.3.0-gpu
I20241008 10:02:22.332681    86 knowhere_config.cc:64] [KNOWHERE][SetSimdType][milvus] FAISS expect simdType::AUTO
I20241008 10:02:22.332715    86 knowhere_config.cc:89] [KNOWHERE][SetSimdType][milvus] FAISS hook AVX2
I20241008 10:02:22.332764    86 Common.cpp:35] [SERVER][SetIndexSliceSize][milvus] set config index slice size (byte): 16777216
I20241008 10:02:22.332840    86 Common.cpp:41] [SERVER][SetHighPriorityThreadCoreCoefficient][milvus] set high priority thread pool core coefficient: 10
I20241008 10:02:22.332866    86 Common.cpp:48] [SERVER][SetMiddlePriorityThreadCoreCoefficient][milvus] set middle priority thread pool core coefficient: 5
I20241008 10:02:22.332886    86 Common.cpp:55] [SERVER][SetLowPriorityThreadCoreCoefficient][milvus] set low priority thread pool core coefficient: 1
I20241008 10:02:22.333035    86 thread_pool.h:152] [KNOWHERE][InitGlobalBuildThreadPool][milvus] Init global build thread pool with size 8
W20241008 10:02:22.335795    86 ExceptionTracer.cpp:187] Invalid trace stack for exception of type: raft::cuda_error
terminate called after throwing an instance of 'raft::cuda_error'
  what():  CUDA error encountered at: file=/root/milvus/cmake_build/thirdparty/knowhere/knowhere-src/src/common/raft/integration/raft_initialization.cc line=53: call='cudaGetDeviceCount(&result)', Reason=cudaErrorInsufficientDriver:CUDA driver version is insufficient for CUDA runtime version
Obtained 7 stack frames
#0 in /opt/lib/libknowhere.so(_ZN4raft9exception18collect_call_stackEv+0x80) [0x7ff758a6ab70]
#1 in /opt/lib/libknowhere.so(_ZN4raft10cuda_errorC1ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0xbd) [0x7ff758a6d30d]
#2 in /opt/lib/libknowhere.so(+0x57bfb1) [0x7ff7588d5fb1]
#3 in /lib64/libc.so.6(+0x8ee18) [0x7ff77df80e18]
#4 in /opt/lib/libknowhere.so(_ZN13raft_knowhere15initialize_raftERKNS_18raft_configurationE+0x56) [0x7ff758dc6626]
#5 in /opt/lib/libknowhere.so(_ZN8knowhere14KnowhereConfig14SetRaftMemPoolEv+0x35) [0x7ff758a34945]
#6 in /opt/milvus/bin/milvus(runtime.asmcgocall.abi0+0x64) [0x1b861c4]

SIGABRT: abort
PC=0x7ff77df7d94c m=13 sigcode=18446744073709551610
signal arrived during cgo execution
```

This does not happen when the cluster has GPU drivers installed. Milvus version used is 2.4.0

**Reply by xiaofan-luan** [reply]
understand. I think milvus failed to init without cuda driver. 
@Presburger 

could we add an error handling to skip this error?

**Comment by presburger** [comment]
@Ald392 Welcome to using the Milvus GPU version. Could you let us know which GPU you are using and the version of the NVIDIA driver? This seems to be an error caused by a mismatch between CUDA and the driver. Milvus GPU requires calling the CUDA runtime API to verify the GPU environment, and if there is an issue with the environment itself, it will result in an error.

**Reply by Ald392** [reply]
My cluster doesn't have a GPU nor a NVIDIA driver. I just want to run the milvus gpu image on a cluster without gpu and it shouldn't provide me an error unless i use any GPU specific indexing.

**Comment by ashkrisk** [comment]
Interestingly, it looks like a Milvus GPU binary built from the master branch can tolerate running on non-GPU machines, but a Milvus GPU binary built from the 2.4.12 branch crashes, complaining about CUDA drivers.

Is there any commit / fix for this in the master branch that's missing from 2.4.x?

**Reply by xiaofan-luan** [reply]
@Presburger 

if there a patch merged to fix GPU image in 2.4.x?

**Reply by ashkrisk** [reply]
Looks like this commit on master is what apparently fixes the issue: https://github.com/milvus-io/milvus/commit/a9352a073b1692f91d7f1c109dbdb8e47cae1afb.

But it's not exactly a fix, it just so happens that it causes Milvus to avoid the code path in Knowhere that triggers the crash. More detailed analysis in the issue I opened: https://github.com/milvus-io/milvus/issues/36894#issuecomment-2414461013

Also raised a Knowhere PR which fixes it: https://github.com/zilliztech/knowhere/pull/892

**Reply by presburger** [reply]
@ashkrisk 
The Milvus GPU memory pool cannot be set to `null`, as this would result in significant memory wastage.
Have you tried adjusting the GPU memory settings in the `milvus.yaml` file for version 2.4.x? Perhaps modifying the YAML parameters could resolve your issue.

**Reply by presburger** [reply]
> @Presburger
> 
> if there a patch merged to fix GPU image in 2.4.x?

I'll submit a PR to fix it right away.
